swaywm/sway
# Basic split containers

Todo:

* [x] basic split containers
* [x] move focus around in the split tree
* [x] kill split containers
* [x] handle container destruction in the seat
* [x] arrange windows properly on container destruction
* [ ] implement organization plan
* [ ] flatten the tree correctly on tree structure changes
* [ ] don't destroy view containers until their surface goes away
* [ ] unify container destroy functions

Organization plan:

* confine all tree structure (attach/move/remove) changes to tree.h
* confine all layout changes (width/height/x/y) changes to layout.h
* confine all traversals, queries, create, and destroy to container.h
* move view and workspaces out of the tree folder

Known bugs: (will be fixed in the next effort)

* Some places in the code destroy the container immediately instead of waiting for the view to close.

Tests:

1. Open two windows and split the right one. Open a new window and you should see it split. Do that for `splith` and `splitv`.
2. Test `focus <direction>` descends into the focus stack. Also test `focus parent` and `focus child` works.
3. Split a container and try to kill it and sway should not crash.
4. destroy a split container with focus and focus should revert to the focus inactive of the first nonempty parent up to the workspace level
5. Open two windows and split the one on the right. Then close the window on the right. The windows should be arranged normally, with the left one taking up the whole screen.
